### ISSUE TITLE
Serializing pre-built automaton

### ISSUE THREAD

**Issue by plusvic** [BurntSushi/aho-corasick#105]
I'm evaluating the possibility of using this crate for a project (https://github.com/VirusTotal/yara-x) as this implementation of Aho-Corasick seems to have all the features I need, and it's very fast. 

The only feature I'm missing so far is the possibility of serializing an automaton, allowing me to load it again from a file without incurring the cost of building the automaton again. The automatons I plan to build are very large, with hundreds of thousands short patterns, even millions. As I don't know much about the internal design of this crate I'm not sure whether a feature like this makes sense, maybe the cost of deserializing the automaton would be similar to the with the cost of building it from scratch. Intuitively, I would say that for small automatons this won't make any differences, but for very large ones deserializing a very large automaton should be faster than building it again. 

I would volunteer for implementing this feature if you think that this is something worth to have, but I would like to hear your advise on this. Thanks for your great work!

**Comment by plusvic** [comment]
Thank you for the detailed response. Now I have a better picture of how this crate works internally, probably re-building the automaton is good enough for my use case. I'll try to experiment a little bit with real data and figure out how big my automatons are going to be.

To be honest I wasn't thinking about zero-copy deserialization, as I was already guessing that it would be hard to implement and maintain. But given how fast the automatons are built, now I doubt that the deserialization could be faster than re-building the automaton, unless we aim for zero-copy deserialization.

**Comment by BurntSushi** [comment]
Note that in #106, I've changed the contiguous NFA to increase the somewhat restricted limit on its total size. Its memory efficiency got a little worse, but now I can build a contiguous NFA from ~15,700,000 Wikipedia titles in about 20 seconds (using 2GB of heap). Previously, I wasn't even able to build one from 1,000,000 Wikipedia titles.

I realize this isn't directly related to serialization, but I just wanted to correct the record above.

**Comment by plusvic** [comment]
This is a bit out of topic, and you may already know about representing the Aho-Corasick automaton as an interleaved state-transition matrix, but I think you may find this article interesting: http://goo.gl/lE6zG. That's the technique I used in the C implementation of YARA, the project that I'm now moving to Rust. 

This comment also explains the technique briefly:
https://github.com/VirusTotal/yara/blob/fa5682a98d9421dfd7a1c35370fb7ff2f89ee7ed/libyara/ahocorasick.c#L469-L524

**Comment by BurntSushi** [comment]
@plusvic Yup! I remember reading that a long time ago. I don't remember why I didn't try it earlier, but the contiguous NFA uses more "classic" finite state machine compression tricks, a lot of which I learned as part of work on the [`fst`](https://github.com/BurntSushi/fst) crate. The trick is balancing compression with the extra overhead each thing requires at search time. One thing that sticks out to my in your link is this:

> // Each 32-bit slot in the transition table has 23 bits for storing the index
> // of the target state and 9 bits for storing the offset of the slot relative
> // to its own state.

Which is kind of the thing I just fixed in the contiguous NFA.

It would be interesting to bake off that approach with my approach in the contiguous NFA. And the new crate structure (on master) I think makes it much easier to add new Aho-Corasick implementations.
